packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -1081,31 +1140,27 @@ function simplifyStruct<T>(typeDef: StructType, context: SimplifyContext<T>): Si
 				: field
 			addField(structKey, mappedField)
 		} else {
-			// TODO: potential optimization: merge unions first before simplifying them,
-			// to avoid having to simplify shared nested spread types multiple times
-			const simplifiedSpreadType = simplify(field.type, {
+			const simplifiedSpread = simplify(field.type, {
 				...context,
 				isMember: true,
 				typeArgs: [],
+				structFields: { literalFields, complexFields },
 			})
-			if (simplifiedSpreadType.kind === 'struct') {
-				for (const field of simplifiedSpreadType.fields) {
-					addField(field.key, field)
-				}
-			} else if (simplifiedSpreadType.kind === 'union') {
-				for (const member of simplifiedSpreadType.members) {
-					if (member.kind === 'struct') {
-						for (const field of member.fields) {
-							// TODO: technically, if a field is required in all members of
-							// this union, it could be added as a required field
-							addField(field.key, { ...field, optional: true })
-						}
-					}
-				}
+			// Recursive calls above will already modify literal and complex fields passed by the
+			// context, so no need to handle them here.
+
+			// In case we are spreading sth that resolves to `any`, allow any other additional field.
+			if (simplifiedSpread.kind === 'any') {
+				addField({ kind: 'any' }, { kind: 'pair', key: { kind: 'any' }, type: { kind: 'any' } })
 			}
-			// Silently ignore spreading non-struct types
 		}
 	}
+	if (context.structFields) {
+		// In this case we are spreading a struct and the fields have been added to the parent map
+		// from context now.
+		return { kind: 'struct', fields: [] }
```

Comment:
Yeah, I also kind of dislike this. I am also not sure how this would work with caching. We would definitely cache the intermediate result of a reference, even when used as a spread operator, which isn't really possible with this anymore.

Reply:
Actually, now that I think about it, the whole deduplicating is likely not needed anymore. The shallow check already constructs a map of literal keys of the runtime value, so if multiple keys in the simplified struct type match, I think it would just overwrite the definition multiple times, keeping the last one. I'll check that after work, that'd probably the best case, both for code abd for performance.

The only case I am not sure about is complex keys